### ISSUE TITLE
LLaMA fails since 0fd44259 on M2

### ISSUE THREAD

**Issue by eichenroth** [tinygrad/tinygrad#2775]
With the commit `0fd44259` (#2698) the LLaMA example broke.

command: `python examples/llama.py --prompt "Hello." --count 10 --temperature 0 --timing`

output:
```
using METAL backend
using LLaMA-7B model
ram used:  0.00 GB, layers.0.attention.wq.weight                      :   0%|                                                                                          | 0/292 [00:00<?, ?it/s]
loaded weights in 48.52 ms, 0.03 GB loaded at 0.69 GB/s
Traceback (most recent call last):
  File "/.../tinygrad/examples/llama.py", line 373, in <module>
    llama = LLaMa.build(MODEL_PATH, TOKENIZER_PATH, model_gen=args.gen, model_size=args.size, quantize=args.quantize)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/.../tinygrad/examples/llama.py", line 174, in build
    load_state_dict(model, weights, strict=False)
  File "/.../tinygrad/tinygrad/nn/state.py", line 68, in load_state_dict
    v.assign(state_dict[k].to(v.device)).realize()
  File "/.../tinygrad/tinygrad/tensor.py", line 105, in realize
    run_schedule(self.lazydata.schedule())
  File "/.../tinygrad/tinygrad/realize.py", line 28, in run_schedule
    prg = lower_schedule_item(si)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/.../tinygrad/tinygrad/realize.py", line 19, in lower_schedule_item
    return Device[si.out.device].get_runner(si.ast)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/.../tinygrad/tinygrad/device.py", line 314, in get_runner
    def get_runner(self, ast:LazyOp) -> CompiledASTRunner: return self.to_program(self.get_linearizer(ast))
                                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/.../tinygrad/tinygrad/device.py", line 286, in to_program
    return CompiledASTRunner(k.ast, k.name, src, k.global_size, k.local_size, runtime_args).build(self.compiler, self.runtime)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/.../tinygrad/tinygrad/device.py", line 254, in build
    self.lib = compiler.__wrapped__(self.prg) if getenv("DISABLE_COMPILER_CACHE") else compiler(self.prg)
                                                                                       ^^^^^^^^^^^^^^^^^^
  File "/.../tinygrad/tinygrad/helpers.py", line 260, in wrapper
    return diskcache_put(table, key, func(*args, **kwargs))
                                     ^^^^^^^^^^^^^^^^^^^^^
  File "/.../tinygrad/tinygrad/runtime/ops_metal.py", line 18, in compile_metal
    library = unwrap2(MetalDevice.compiler_device.newLibraryWithSource_options_error_(prg, options, None))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/.../tinygrad/tinygrad/helpers.py", line 45, in unwrap2
    assert err is None, str(err)
AssertionError: Error Domain=MTLLibraryErrorDomain Code=3 "program_source:3:55: error: invalid type 'const device __bf16 *' for buffer declaration
kernel void E_131072_32_4(device unsigned int* data0, const device __bf16* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
program_source:3:68: note: type '__bf16' cannot be used in buffer pointee type
kernel void E_131072_32_4(device unsigned int* data0, const device __bf16* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
                                                                   ^
" UserInfo={NSLocalizedDescription=program_source:3:55: error: invalid type 'const device __bf16 *' for buffer declaration
kernel void E_131072_32_4(device unsigned int* data0, const device __bf16* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
program_source:3:68: note: type '__bf16' cannot be used in buffer pointee type
kernel void E_131072_32_4(device unsigned int* data0, const device __bf16* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
                                                                   ^
}
```

**Comment by chenyuxyz** [comment]
i think this has been fixed on master. can you verify?

**Comment by eichenroth** [comment]
Not fixed unfortunately.

**Comment by chenyuxyz** [comment]
what's the error you are getting? is it the same as https://github.com/tinygrad/tinygrad/issues/2226? or still something related to bf16?

**Comment by soitgoes** [comment]
+1. M2 broken for coder.py starting 0fd44259

looking at the diff, I guess M2 doesn't support bfloat16.  Over my ski's here.  Hope this helps.

create model: 145.43 ms
download weights: 75.65 ms
ram used:  0.00 GB, layers.0.attention.wq.weight                      :   0%| | 
loaded weights in 75.73 ms, 0.03 GB loaded at 0.44 GB/s
weights -> model: 111.40 ms
Traceback (most recent call last):
  File "/Users/martin/working/tinygrad/examples/coder.py", line 37, in <module>
    nn.state.load_state_dict(model, fix_bf16(convert_from_huggingface(part1, model, 32, 8)), strict=False)
  File "/Users/martin/working/tinygrad/tinygrad/nn/state.py", line 60, in load_state_dict
    v.assign(state_dict[k].to(v.device)).realize()
  File "/Users/martin/working/tinygrad/tinygrad/tensor.py", line 105, in realize
    run_schedule(self.lazydata.schedule())
  File "/Users/martin/working/tinygrad/tinygrad/realize.py", line 28, in run_schedule
    prg = lower_schedule_item(si)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/working/tinygrad/tinygrad/realize.py", line 19, in lower_schedule_item
    return Device[si.out.device].get_runner(si.ast)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/working/tinygrad/tinygrad/device.py", line 308, in get_runner
    def get_runner(self, ast:LazyOp) -> CompiledASTRunner: return self.to_program(self.get_linearizer(ast))
                                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/working/tinygrad/tinygrad/device.py", line 280, in to_program
    return CompiledASTRunner(k.ast, k.name, src, k.global_size, k.local_size, runtime_args).build(self.compiler, self.runtime)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/working/tinygrad/tinygrad/device.py", line 249, in build
    self.lib = compiler.__wrapped__(self.prg) if getenv("DISABLE_COMPILER_CACHE") else compiler(self.prg)
                                                                                       ^^^^^^^^^^^^^^^^^^
  File "/Users/martin/working/tinygrad/tinygrad/helpers.py", line 240, in wrapper
    return diskcache_put(table, key, func(*args, **kwargs))
                                     ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/working/tinygrad/tinygrad/runtime/ops_metal.py", line 18, in compile_metal
    library = unwrap2(MetalDevice.compiler_device.newLibraryWithSource_options_error_(prg, options, None))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/martin/working/tinygrad/tinygrad/helpers.py", line 45, in unwrap2
    assert err is None, str(err)
           ^^^^^^^^^^^
AssertionError: Error Domain=MTLLibraryErrorDomain Code=3 "program_source:3:57: error: invalid type 'const device __bf16 *' for buffer declaration
kernel void E_512_64_8_16_4(device unsigned int* data0, const device __bf16* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
program_source:3:70: note: type '__bf16' cannot be used in buffer pointee type
kernel void E_512_64_8_16_4(device unsigned int* data0, const device __bf16* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
                                                                     ^
" UserInfo={NSLocalizedDescription=program_source:3:57: error: invalid type 'const device __bf16 *' for buffer declaration
kernel void E_512_64_8_16_4(device unsigned int* data0, const device __bf16* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
program_source:3:70: note: type '__bf16' cannot be used in buffer pointee type
kernel void E_512_64_8_16_4(device unsigned int* data0, const device __bf16* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
                                                                     ^
}

**Comment by eichenroth** [comment]
> what's the error you are getting? is it the same as #2226? or still something related to bf16?

no its not the same issue, it seems to be the issue the bf16-cast as @soitgoes mentioned.

**Comment by chenyuxyz** [comment]
please reopen if #3774 does not fix the issue for you

**Comment by eichenroth** [comment]
@chenyuxyz seems not to be fixed. Is some of the C-code cached before compilation? Happy to check again

**Comment by chenyuxyz** [comment]
oh i only fixed coder in my earlier, missed that llama has the treatment. should be fixed after #3789. If your device does not support bfloat16 buffer, you need to add SUPPORT_BF16=0 in the command

**Comment by eichenroth** [comment]
Indeed fixed on my Mac using the command: `SUPPORT_BF16=0 python examples/llama.py --prompt "Hello."`
